cumulus/client/consensus/aura/src/collators/slot_based/block_builder_task.rs

```diff
@@ -0,0 +1,390 @@
+		let timestamp = sp_timestamp::Timestamp::current();
+		SlotInfo { slot: Slot::from_timestamp(timestamp, slot_duration), timestamp, slot_duration }
+	}
+}
+
+/// Run block-builder.
+pub async fn run_block_builder<Block, P, BI, CIDP, Client, Backend, RClient, CHP, Proposer, CS>(
+	params: BuilderTaskParams<Block, BI, CIDP, Client, Backend, RClient, CHP, Proposer, CS>,
+) where
+	Block: BlockT,
+	Client: ProvideRuntimeApi<Block>
+		+ UsageProvider<Block>
+		+ BlockOf
+		+ AuxStore
+		+ HeaderBackend<Block>
+		+ BlockBackend<Block>
+		+ Send
+		+ Sync
+		+ 'static,
+	Client::Api:
+		AuraApi<Block, P::Public> + CollectCollationInfo<Block> + AuraUnincludedSegmentApi<Block>,
+	Backend: sc_client_api::Backend<Block> + 'static,
+	RClient: RelayChainInterface + Clone + 'static,
+	CIDP: CreateInherentDataProviders<Block, ()> + 'static,
+	CIDP::InherentDataProviders: Send,
+	BI: BlockImport<Block> + ParachainBlockImportMarker + Send + Sync + 'static,
+	Proposer: ProposerInterface<Block> + Send + Sync + 'static,
+	CS: CollatorServiceInterface<Block> + Send + Sync + 'static,
+	CHP: consensus_common::ValidationCodeHashProvider<Block::Hash> + Send + 'static,
+	P: Pair,
+	P::Public: AppPublic + Member + Codec,
+	P::Signature: TryFrom<Vec<u8>> + Member + Codec,
+{
+	let BuilderTaskParams {
+		relay_client,
+		create_inherent_data_providers,
+		para_client,
+		keystore,
+		block_import,
+		para_id,
+		proposer,
+		collator_service,
+		collator_sender,
+		code_hash_provider,
+		authoring_duration,
+		para_backend,
+		relay_chain_slot_duration,
+		slot_drift,
+	} = params;
+
+	let slot_timer = SlotTimer::<_, _, P>::new_with_drift(para_client.clone(), slot_drift);
+
+	let mut collator = {
+		let params = collator_util::Params {
+			create_inherent_data_providers,
+			block_import,
+			relay_client: relay_client.clone(),
+			keystore: keystore.clone(),
+			para_id,
+			proposer,
+			collator_service,
+		};
+
+		collator_util::Collator::<Block, P, _, _, _, _, _>::new(params)
+	};
+
+	loop {
+		// We wait here until the next slot arrives.
+		let para_slot = slot_timer.wait_until_next_slot().await;
+
+		let Ok(expected_cores) =
+			expected_core_count(relay_chain_slot_duration, para_slot.slot_duration)
+		else {
+			return
+		};
+
+		let Ok(relay_parent) = relay_client.best_block_hash().await else {
+			tracing::warn!("Unable to fetch latest relay chain block hash, skipping slot.");
+			continue;
+		};
+
+		let scheduled_cores = cores_scheduled_for_para(relay_parent, para_id, &relay_client).await;
+		if scheduled_cores.is_empty() {
+			tracing::debug!(target: LOG_TARGET, "Parachain not scheduled, skipping slot.");
+			continue;
+		}
+
+		let core_index_in_scheduled: u64 = *para_slot.slot % expected_cores;
+		let Some(core_index) = scheduled_cores.get(core_index_in_scheduled as usize) else {
+			tracing::debug!(target: LOG_TARGET, core_index_in_scheduled, core_len = scheduled_cores.len(), "Para is scheduled, but not enough cores available.");
```

Comment:
For fixed scaling might make sense to have this as a warning, so we can act on it.